src/pool/srv_cli.c

```diff
@@ -491,3 +497,615 @@ dsc_pool_svc_query(uuid_t pool_uuid, d_rank_list_t *ps_ranks, uint64_t deadline,
+	int                           i;
+	int                           rc = out->pqio_op.po_rc;
+
+	if (rc != 0) {
+		DL_ERROR(rc, DF_UUID ": failed to query pool rank %u target %u", DP_UUID(pool_uuid),
+			 arg->pqta_rank, arg->pqta_tgt_idx);
+		return rc;
+	}
+
+	D_DEBUG(DB_MGMT, DF_UUID ": Successfully queried pool rank %u target %u\n",
+		DP_UUID(pool_uuid), arg->pqta_rank, arg->pqta_tgt_idx);
+
+	arg->pqta_info->ta_type  = DAOS_TP_UNKNOWN;
+	arg->pqta_info->ta_state = out->pqio_state;
+	for (i = 0; i < DAOS_MEDIA_MAX; i++) {
+		arg->pqta_info->ta_space.s_total[i] = out->pqio_space.s_total[i];
+		arg->pqta_info->ta_space.s_free[i]  = out->pqio_space.s_free[i];
+	}
+
+	return 0;
+}
+
+static struct dsc_pool_svc_call_cbs pool_query_target_cbs = {
+	.pscc_op	= POOL_QUERY_INFO,
+	.pscc_init	= pool_query_target_init,
+	.pscc_consume	= pool_query_target_consume,
+	.pscc_fini	= NULL
+};
+
+/**
+ * Query pool target information without holding a pool handle.
+ *
+ * \param[in]	pool_uuid	UUID of the pool
+ * \param[in]	ps_ranks	Ranks of pool svc replicas
+ * \param[in]	deadline	Unix time deadline in milliseconds
+ * \param[in]	rank		Pool storage engine rank
+ * \param[in]	tgt_idx		Target index within the pool storage engine
+ * \param[out]	ti		Target information (state, storage capacity and usage)
+ *
+ * \return	0		Success
+ *		-DER_INVAL	Invalid input
+ *		Negative value	Other error
+ */
+int
+dsc_pool_svc_query_target(uuid_t pool_uuid, d_rank_list_t *ps_ranks, uint64_t deadline,
+			  d_rank_t rank, uint32_t tgt_idx, daos_target_info_t *ti)
+{
+	struct pool_query_target_arg arg = {
+		.pqta_rank	= rank,
+		.pqta_tgt_idx	= tgt_idx,
+		.pqta_info	= ti
+	};
+
+	if (ti == NULL)
+		return -DER_INVAL;
+	D_DEBUG(DB_MGMT, DF_UUID ": Querying pool target %u\n", DP_UUID(pool_uuid), tgt_idx);
+	return dsc_pool_svc_call(pool_uuid, ps_ranks, &pool_query_target_cbs, &arg, deadline);
+}
+
+struct pool_evict_arg {
+	uuid_t   *pea_handles;
+	size_t    pea_n_handles;
+	char     *pea_machine;
+	uint32_t  pea_destroy;
+	uint32_t  pea_force;
+	uint32_t *pea_count;
+};
+
+static int
+pool_evict_init(uuid_t pool_uuid, crt_rpc_t *rpc, void *varg)
+{
+	struct pool_evict_arg *arg = varg;
+	struct pool_evict_in  *in  = crt_req_get(rpc);
+
+	in->pvi_hdls.ca_arrays = arg->pea_handles;
+	in->pvi_hdls.ca_count  = arg->pea_n_handles;
+	in->pvi_machine        = arg->pea_machine;
+	/* Pool destroy (force=false): assert no open handles / do not evict.
+	 * Pool destroy (force=true): evict any/all open handles on the pool.
+	 */
+	in->pvi_pool_destroy       = arg->pea_destroy;
+	in->pvi_pool_destroy_force = arg->pea_force;
+	return 0;
+}
+
+static int
+pool_evict_consume(uuid_t pool_uuid, crt_rpc_t *rpc, void *varg)
+{
+	struct pool_evict_arg *arg = varg;
+	struct pool_evict_out *out = crt_reply_get(rpc);
+	int                    rc  = out->pvo_op.po_rc;
+
+	if (rc != 0)
+		DL_ERROR(rc, DF_UUID ": pool destroy failed to evict handles", DP_UUID(pool_uuid));
```

Comment:
should this return rc in the error case, rather than proceed to set *arg->pea_count?
I can't remember if the failure execution flow still requires the value to be set.